### PR TITLE
Default member access inspections

### LINKED ISSUE #2504: Inspection for implicit default member usage on RHS
Following two Excel-specific examples where implicit calls to `Range.[_Default]` result in unexpected outcomes (where *some* values of unexpected involve the host application exploding into a thousand tiny pieces), it might be time to add an inspection for implicit default members on the RHS.

- [Variant array is corrupted](http://stackoverflow.com/questions/34951023/variant-array-is-corrupted-when-running-macro-excel-crashes/41539288#41539288) when assigned an implicit `Range.[_Default]` 

- [Cell copy character limit](http://stackoverflow.com/questions/40314695/vba-cell-copy-character-limit/40315304#40315304)

In both instances above, the problems are alleviated by explicitly calling the `Range.[_Default]` member (or its equivalent `Range.Value` member).

But there are *lots* of objects that have default members, and not all of the default members are necessarily visible or descriptive by default. 

- `Excel.Application.[_Default]` / `Excel.Application.Name`
- `Range.[_Default]` / `Excel.Range.Value`
- `Excel.Shapes.[_Default]` / `Excel.Shapes.Item`
- `ADODB.Connection.ConnectionString`
- `ADODB.Recordset.Fields`
- `ADODB.Field.Value`

I'm unsure how best to implement this inspection. There are a number of approaches that could be taken:

1. Find/Fix *all* implicit member usage. This might be overkill, as the number of resulting explicit usages is likely to be quite high

2. Find/Fix *all* implicit, Variant-returning members. The RHS problems all seem to be related to members returning Variants. Maybe this inspection should be specific to implicit default members that return Variants?

3. Find/Fix a user-configurable list of implicit members. Allow the use to define `Excel.Range` as the implicit member, and define `Excel.Range.Value` as the preferred/overridden explicit member. I'd certainly rather see `myRange.Value` in my code than `myRange.[_Default]`, and `Sheets.Item("foo")` over `Sheets.[_Default]("foo")`

Maybe the RD settings could allow for any one of the above to be configured?

### LINKED ISSUE #5093: Refactor bang operators into dot operators
**Justification**
When replying to #5092 I thought we already had a request for this refactoring but was surprised that there wasn't. 

The proposal is to have a simple refactor that converts all bang operators into dot operators -- there is no legit reason to use bang operators, especially on `Me` on the LHS. The refactor should expand the expressions (e.g. Access' form's `Me!Foo` should become `Me.Foo` which is more descriptive of what the code is actually doing and more importantly makes the code verifiable. 

An unverifiable code is unrefactorable code. Rubberduck should help them write verifiable code.

**Description**
Given:

Assuming an Access form and a DAO Recordset:
```
Me!Foo = rs!Bar 
```

The outcome should be:
```
Me.Foo.Value = rs.Fields("Bar").Value
```

Note that the Access form can have dot access to the controls without going through the `Me.Controls` collection, but we can't do that for the recordset. That may require some inspection on the object to determine whether we can dot access the member or use the default collection.

### PR DESCRIPTION
**Pull request by MDoerner**
Closes #2504 
Closes #5093 

This PR reorganizes and extends the existing default member inspections. 
It introduces a set of three inspections for indexed and implicit default membre accesses each. Both times one for unboud accesses, one for those requiring a recursive default member resolution and the remaining ordinary ones. Inexchange, the `ImplicitDefaultMemberAssignmentInspection` is removed as it is superseded by the other inspections.

In addition to the extensionand reorganization, tests are added to verify that the `ExpandDefaultMemberQuickFix` can fix all the bound inspection results. (For the unbound ones, we lack the necessary information.)

NOTE:
Because this PR uses the quick fix from PR #5165, it is based on that PR. Please do not merge this PR before it. I will rebase this PR once the other one is merged.

### CODE REVIEW

File: Rubberduck.CodeAnalysis/Inspections/Concrete/ImplicitDefaultMemberAccessInspection.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+using Rubberduck.Resources.Inspections;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Inspections.Inspections.Extensions;
+using Rubberduck.Parsing.Inspections;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    /// <summary>
+    /// Identifies the use non-indexed default member accesses.
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/ImplicitDefaultMemberAccessInspection.cs]: "the use *of* non-indexed..."

File: Rubberduck.CodeAnalysis/Inspections/Concrete/ImplicitRecursiveDefaultMemberAccessInspection.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+using Rubberduck.Resources.Inspections;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Inspections.Inspections.Extensions;
+using Rubberduck.Parsing.Inspections;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    /// <summary>
+    /// Identifies the use indexed default member accesses that require a recursive default member resolution.
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/ImplicitRecursiveDefaultMemberAccessInspection.cs]: "the use *of* indexed..."

File: Rubberduck.CodeAnalysis/Inspections/Concrete/ImplicitUnboundDefaultMemberAccessInspection.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Inspections.Inspections.Extensions;
+using Rubberduck.Parsing.Inspections;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    /// <summary>
+    /// Identifies the use indexed default member accesses for which the default member cannot be determined at compile time.
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/ImplicitUnboundDefaultMemberAccessInspection.cs]: Ditto

File: Rubberduck.CodeAnalysis/Inspections/Concrete/IndexedDefaultMemberAccessInspection.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Inspections.Inspections.Extensions;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Inspections;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    /// <summary>
+    /// Identifies the use indexed default member accesses.
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/IndexedDefaultMemberAccessInspection.cs]: ditto

File: RubberduckTests/Inspections/ImplicitDefaultMemberAccessInspectionTests.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+Attribute Foo.VB_UserMemId = 0
+End Function
+";
+
+            var moduleCode = $@"
+Private Function Foo() As Long 
+    Dim cls As Object
+    Set cls = New Class1
+    Dim bar As Long
+    {statement}
```
Comment (retailcoder) [inline RubberduckTests/Inspections/ImplicitDefaultMemberAccessInspectionTests.cs]: Nice

File: RubberduckTests/Inspections/ImplicitDefaultMemberAccessInspectionTests.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                ("Module1", moduleCode, ComponentType.StandardModule));
+
+            var inspectionResults = InspectionResults(vbe.Object);
+
+            Assert.AreEqual(2, inspectionResults.Count());
+        }
+
+        protected override IInspection InspectionUnderTest(RubberduckParserState state)
+        {
+            return new ImplicitDefaultMemberAccessInspection(state);
```
Comment (retailcoder) [inline RubberduckTests/Inspections/ImplicitDefaultMemberAccessInspectionTests.cs]: Could be expression-bodied (nit)

File: Rubberduck.CodeAnalysis/Inspections/Concrete/IndexedDefaultMemberAccessInspection.cs (hunk 4 above)
Comment (MDoerner) [reply Rubberduck.CodeAnalysis/Inspections/Concrete/IndexedDefaultMemberAccessInspection.cs]: The copy-pasta is real.

File: RubberduckTests/Inspections/ImplicitDefaultMemberAccessInspectionTests.cs (hunk 6 above)
Comment (MDoerner) [reply RubberduckTests/Inspections/ImplicitDefaultMemberAccessInspectionTests.cs]: TBH, I prefer to use expression bodied members only for properties. If you want me to turn this into an expression bodied meber, just poke me again.

**Pull request merged**